### PR TITLE
Add support for the SPV_QCOM_image_processing extension

### PR DESCRIPTION
**Pull request by wooyoungqcom**
1. Add support for the SPV_QCOM_image_processing extension
2. Add four unit tests
3. Fix minor issues in other unit tests

Please note that 

the following test failed with "The flag --preserve-numeric-ids was specified multiple times."

shaders-no-opt/asm/frag/opaque-id-literal-alias.preserve.asm.frag

the following test failed with "The type of Value must match the Result type"

shaders-no-opt/asm/frag/subgroup-arithmetic-cast.nocompat.vk.asm.frag
shaders-msl-no-opt/asm/frag/subgroup-arithmetic-cast.msl21.asm.frag
  %18 = OpGroupNonUniformSMin %uint %uint_3 Reduce %16

the following tests failed with "Selection must be structured"

shaders-no-opt/asm/comp/loop-variable-early-read-with-initializer.asm.comp
  OpBranchConditional %69 %12 %13

shaders-no-opt/asm/comp/loop-variable-early-read-with-undef.asm.comp
  OpBranchConditional %69 %12 %13

the following tests failed with "OpSpecConstantComposite Constituent <id> '7[%7]'s type does not match Result Type <id> '29[%v3uint]'s vector element type."

shaders-no-opt/asm/comp/local-size-id.vk.asm.comp
shaders-hlsl-no-opt/asm/comp/local-size-id.asm.comp
shaders-msl-no-opt/asm/comp/local-size-id.asm.comp
  %38 = OpSpecConstantComposite %v3uint %37 %7 %uint_2

	
	shaders-msl-no-opt/asm/comp/local-size-id.asm.comp
	shaders-msl-no-opt/asm/frag/subgroup-arithmetic-cast.msl21.asm.frag

### CODE REVIEW

**Comment by HansKristian-Work** [comment]
Rolling deps separately in https://github.com/KhronosGroup/SPIRV-Cross/pull/2254. You can rebase against that when it lands.

**Comment by wooyoungqcom** [comment]
Rebased the branch.

File: spirv_glsl.cpp [hunk 1]
```diff
@@ -14148,6 +14148,66 @@ void CompilerGLSL::emit_instruction(const Instruction &instruction)
...
+	case OpImageBlockMatchSADQCOM:
+	{
+		require_extension_internal("GL_QCOM_image_processing");
+		uint32_t result_type_id = ops[0];
+		uint32_t id = ops[1];
+		string expr;
+		switch (opcode)
+		{
+			case OpImageSampleWeightedQCOM:
+				expr = std::move("textureWeightedQCOM");
```
Comment (HansKristian-Work) [inline spirv_glsl.cpp:14163]:
```suggestion
				expr = "textureWeightedQCOM";
```

File: spirv_glsl.cpp [hunk 2]
```diff
@@ -14148,6 +14148,66 @@ void CompilerGLSL::emit_instruction(const Instruction &instruction)
...
+				break;
+			case OpImageBoxFilterQCOM:
+				expr = std::move("textureBoxFilterQCOM");
```
Comment (HansKristian-Work) [inline spirv_glsl.cpp:14166]:
```suggestion
				expr = "textureBoxFilterQCOM";
```

File: spirv_glsl.cpp [hunk 3]
```diff
@@ -14148,6 +14148,66 @@ void CompilerGLSL::emit_instruction(const Instruction &instruction)
...
+				break;
+			case OpImageBlockMatchSSDQCOM:
+				expr = std::move("textureBlockMatchSSDQCOM");
```
Comment (HansKristian-Work) [inline spirv_glsl.cpp:14169]:
```suggestion
				expr = "textureBlockMatchSSDQCOM";
```

File: spirv_glsl.cpp [hunk 4]
```diff
@@ -14148,6 +14148,66 @@ void CompilerGLSL::emit_instruction(const Instruction &instruction)
...
+				break;
+			case OpImageBlockMatchSADQCOM:
+				expr = std::move("textureBlockMatchSADQCOM");
```
Comment (HansKristian-Work) [inline spirv_glsl.cpp:14172]:
```suggestion
				expr = "textureBlockMatchSADQCOM";
```

File: spirv_glsl.cpp [hunk 5]
```diff
@@ -14148,6 +14148,66 @@ void CompilerGLSL::emit_instruction(const Instruction &instruction)
...
+				break;
+			default:
+				SPIRV_CROSS_THROW("Invalid opcode for QCOM_image_processing.");
+		}
+
+		expr += ")";
+		emit_op(result_type_id, id, expr, forward);
+
+		inherit_expression_dependencies(id, ops[3]);
+		if (opcode == OpImageBlockMatchSSDQCOM || opcode == OpImageBlockMatchSADQCOM )
```
Comment (HansKristian-Work) [inline spirv_glsl.cpp:14205]:
```suggestion
		if (opcode == OpImageBlockMatchSSDQCOM || opcode == OpImageBlockMatchSADQCOM)
```

File: spirv_glsl.cpp [hunk 6]
```diff
@@ -14148,6 +14148,66 @@ void CompilerGLSL::emit_instruction(const Instruction &instruction)
...
+				expr += ", " + to_expression(ops[4]);
+				break;
+			case OpImageBlockMatchSSDQCOM:
+			case OpImageBlockMatchSADQCOM:
+				expr += ", " + to_non_uniform_aware_expression(ops[4]);
+				expr += ", " + to_expression(ops[5]);
+				expr += ", " + to_expression(ops[6]);
+				break;
+			default:
+				SPIRV_CROSS_THROW("Invalid opcode for QCOM_image_processing.");
```
Comment (HansKristian-Work) [inline spirv_glsl.cpp:14198]:
```suggestion
		case OpImageSampleWeightedQCOM:
			expr += ", " + to_non_uniform_aware_expression(ops[4]);
			break;
		case OpImageBoxFilterQCOM:
			expr += ", " + to_expression(ops[4]);
			break;
		case OpImageBlockMatchSSDQCOM:
		case OpImageBlockMatchSADQCOM:
			expr += ", " + to_non_uniform_aware_expression(ops[4]);
			expr += ", " + to_expression(ops[5]);
			expr += ", " + to_expression(ops[6]);
			break;
		default:
			SPIRV_CROSS_THROW("Invalid opcode for QCOM_image_processing.");
```

**Review by HansKristian-Work — CHANGES_REQUESTED** [review]

**Comment by wooyoungqcom** [comment]
I have made the changes requested.

**Pull request merged**